src/pages/Chat/Chat/index.tsx

```diff
@@ -11,153 +15,110 @@ import MaterialIcons from 'react-native-vector-icons/MaterialIcons';
+
+  const sorter = (a: IMessage, b: IMessage): number => (b.createdAt as Date).getTime() - (a.createdAt as Date).getTime();
+
+  const loadMessages = useCallback((lastMessage?: IMessage): void => {
+    if (!chatRef) return;
+
+    const lastTimestamp = lastMessage ? lastMessage.createdAt as Date : undefined;
+
+    chatAPI.loadMessages(chatRef, 10, lastTimestamp)
+      .then((loadedMessages) => {
+        const chatMessages = loadedMessages.docs.map((message) => {
+          const messageData = message.data();
+          return {
+            _id: message.id,
+            text: messageData.text,
+            createdAt: messageData.timestamp ? messageData.timestamp.toDate() : new Date(),
+            user: {
+              _id: messageData.sender.id,
+            },
+          };
+        });
+
+        setMessages((previousMessages) => GiftedChat.append(previousMessages, chatMessages).sort(sorter));
+
+        setIsLoadingEarlier(false);
+      })
+      .catch(() => {
+        Alert.alert(
+          'Erro!',
+          'Ocorreu um erro no carregamento das informações.\n'
+              + 'Retornando para a página anterior.',
+          [
+            {
+              text: 'Ok',
+              onPress: () => navigation.goBack(),
+            },
+          ],
+          {
+            cancelable: false,
+          },
+        );
+      });
+  }, [chatRef, navigation]);
 
   useEffect(() => {
-    // eslint-disable-next-line
-    console.log('================ CHAT PAGE ================');
-    // eslint-disable-next-line
-    console.log('chatAPI.getChat(currentUser, targetUser, animal): ', targetUserUID, animalUID);
-    // eslint-disable-next-line
-    console.log('...then');
-    // eslint-disable-next-line
-    console.log('chatAPI.loadMessages(chat, pageSize) if chat');
-    setMessages(mock.sort((a, b) => b.createdAt.getTime() - a.createdAt.getTime()));
-  }, [targetUserUID, animalUID]);
-
-  const onSend = useCallback((newMessages = []) => {
-    // eslint-disable-next-line
-    console.log('chatAPI.pushPessages(chat, newMessages): ', newMessages);
-    setMessages((previousMessages) => GiftedChat.append(previousMessages, newMessages));
+    if (routeChatUID !== undefined) {
+      setChatRef(chatAPI.chatDocument(routeChatUID));
+      chatAPI.markChatMessagesAsSeemByUser(routeChatUID, currentUserRef);
+    } else if (routeChatUID === undefined && targetUserUID !== undefined) {
+      chatAPI.getChatByTarget(currentUserRef, userAPI.userDocument(targetUserUID)).then((chat) => {
+        if (chat?.ref) {
+          setChatRef(chat.ref);
+          chatAPI.markChatMessagesAsSeemByUser(chat.ref.id, currentUserRef);
+        } else {
+          setMessages([]);
+        }
+      });
+    }
+  // eslint-disable-next-line react-hooks/exhaustive-deps
   }, []);
 
+  useEffect(() => loadMessages(), [loadMessages, chatRef]);
+
+  const onSend = useCallback(async (newMessages: IMessage[] = []) => {
+    let cRef: DocumentRefData;
```

Comment:
Question: Do we really need cRef? Can't we just setChatRef if chatRef is undefined and then use chatRef assuming it is not undefined?

Reply:
Unfortunatelly no, setChatRef is asynchronous.

Reply:
Ok, but can't you use `await` in an async function?

Reply:
setChatRef is something internal of react. It isn's an asynchronous function but it's result won't have effect immediately.